doc/OVERVIEW.md

```diff
@@ -0,0 +1,79 @@
+
+This library is an implementation of that.
+
+
+## Use dynamic code generator to simulate static code
+
+Let's say we have this very simple class.
+
+```csharp
+```
+
+We can write the following methods to create an object and call the Introduce method. Note that all arguments and return values are cast to `object` type. This is to make sure that the interface of our library is generic enough to handle all method signatures. You might notice that it is the same approach that the ReflectionAPI uses.
+
+```csharp
+```
+
+Given a class name, method name, and the list of parameter types, our goal is to use dynamic code generator to create delegates that are functionally identical to the methods above. If we use the same IL code, the performance of those delegates should be comparable to static code. They can then be used to replace the ReflectionAPI.
+
+
+## The flow to create a delegate
+
+```csharp
+```
+
+The steps are below.
+- Use the provided class name, method name, and the list of parameter types to retrieve the corresponding `MethodInfo` (or `ConstructorInfo` if we are trying to call a constructor).
+- Emit the necessary IL code to load arguments into the stack, execute the method, and return the result (if any).
+- Wrap everything into a delegate.
+
+Moreover, because it is likely that we would want to repeatedly call the same method with different arguments, it makes sense to store the created delegates in a cache. The cache key is crafted from the class name, the method name, and the list of method parameters (generic ones included). By default, each cache entry has a sliding expiration time of 12 hours. But this is configurable.
+
+
+## Non-generic code sample
+
+Below is the code to create objects of type `TestClass`. Note that for a parameterless constructor/method, we can explicitly use `Type.EmptyType` as parameter types, or we can omit it altogether.
+
+```csharp
+```
+
+Then we can call the `Introduce` method on the created object.
+
+```csharp
+```
+
+To call a method that returns void, we simply switch from `Invocator.CreateFunc` to `Invocator.CreateAction`.
+
+
+## Generic code sample
+
+To call a generic method, we need to provide a list of concrete types, as well as specify which method parameters are generic. This information is provided by the `GenericInfo` type.
+
+```csharp
+```
+
+Let's say we want to call a method with this signature: `public void Method<T1, T2>(string s, T2 input)`.
+
+We can see that it has two generic types, `T1` and `T2`. Moreover, its second parameter (index 1) is a generic parameter. To call it with `T1 == int` and `T2 == string`, we create this `GenericInfo` object and pass it to the `CreateFunc` method.
+
+```csharp
+```
+
+For non-generic methods, `GenericInfo` can be omitted.
+
+## Benchmark
+
+This is the comparison between static code, this library, and the ReflectionAPI.
+
+![]()
```

Comment:
should this entire benchmark section be scrapped (README has benchmark results) and if you can re-gen a PNG image of the benchmarks, add it on the README?

Reply:
Okay that makes sense. I'll replace the table on the README with a PNG image.

Reply:
The images were generated from the data in those tables. So I think if we move the images to README then the tables would become redundant.